File: project/app/javascript/channels/war_channel.js
import { App, Helper } from "srcs/internal";
import consumer from "./consumer";

export function ConnectWarChannel(war_id) {
  return consumer.subscriptions.create(
    {
      channel: "WarChannel",
      war_id: war_id,
    },
    {
      connected() {},

      disconnected() {
        this.unsubscribe();
      },

      received(data) {
        if (data.type === "no_reply" && Helper.isCurrentUser(data.user_id)) {
          Helper.info({
            subject: "상대 길드 미응답",
            description: "상대 길드가 응답하지 않아 배틀에서 승리하셨습니다!",
          });
          App.router.navigate(`#/guilds/${App.current_user.get("guild").id}`);
        } else if (Helper.isCurrentView("war-index-view")) {
          this.updateWarBattleView(data);
        }
      },

      updateWarBattleView(data) {
        App.main_view.current_view.war_battle_view?.updateView(data);
      },

      requestBattle(data) {
        this.perform("request_battle", {
          war_id: data.war_id,
          guild_id: data.guild_id,
          match_id: data.match_id,
        });
      },
    }
  );
}

File: project/app/channels/war_channel.rb
class WarChannel < ApplicationCable::Channel
  def subscribed
    stream_from "war_channel_#{params[:war_id].to_s}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # 한 길드의 유저가 전투 버튼을 눌렀을 때에만 호출
  def request_battle(data)
    ActionCable.server.broadcast(
      "war_channel_#{data['war_id'].to_s}",
      {
        type: "request",
        guild_id: data["guild_id"],
        match_id: data["match_id"],
      },
    )
  end
end
